### DISCUSSION TITLE
Usage of Botan v3.x along with project implemented with c++14

### DISCUSSION THREAD

**Discussion by omaralam96** [randombit/botan#3631] [Q&A] [answered]
Would be any problems popped up for me during inclusion header of botan or usage of this library along with project has a constraint of usage only C++14 ?
Knowing that I am able to use GCC v12

**Comment by reneme** [comment, marked as answer]
[GCC 12 should be enough to build Botan 3.x](https://botan.randombit.net/handbook/support.html#supported-platforms). Nevertheless, the library uses C++20 features (eg. `std::span` and concepts) in the public headers. If your code base is restricted to C++14 that will likely be a problem, as long as you don't want to wrap things into your own interface or use [Botan's C-Bindings](https://botan.randombit.net/handbook/api_ref/ffi.html).

Anyway, as long as you don't require [new features added in Botan 3.x](https://github.com/randombit/botan/blob/master/news.rst), you could stay on Botan 2.x that is a C++11 code base. Support and security fixes -- if necessary -- will be available [until at least the end of next year](https://botan.randombit.net/handbook/support.html#branch-support-status).

**Reply by omaralam96** [reply]
Unfortunately, I need to work with TLS v1.3 which is a feature of botan 3.x.
according to my understanding, if I intend to work with botan 3.x I need to use Botan's C-Binding, but I am not seeing any api for TLS in FFI. Does FFI have all functionality provided by botan3.x ??

**Reply by reneme** [reply]
You are right. TLS 1.3 is neither part of Botan 2 nor is it currently exposed via the C bindings. Adding it to the bindings would certainly be a nice improvement. Though, I don't have capacity for it right now.

Maybe you could extract the TLS functionality into a small library that is compiled using C++20 but exposes a C++14-compatible interface. What's the TCP networking stack you're planning to use?

**Reply by omaralam96** [reply]
Regard to your question: I am planing to integrate TLS with SomeIP so that I could secure communication across transport layer.

However, you suggest that the only solution I could do so that I could work with botan 3.x is to extract functionality I want to use into an external library and build its own c++14 interface. I am getting you, right ?!

Thank you for your quick reply BTW <3

**Reply by reneme** [reply]
>  I could work with botan 3.x is to extract functionality I want to use into an external library and build its own c++14 interface. I am getting you, right ?!

That's what I mean, yes. Basically write a wrapper around Botan's APIs that expose the functionality you need via an interface that uses C++14 exclusively. You'll need to compile this wrapper using C++20, but you should be okay to include and use your wrapper's header files in a C++14 project.

Disclaimer: Personally I haven't been in a situation that required such an approach. But I don't see why this wouldn't work.

Another option, of course, would be to compile your project in C++20 mode. If you use GCC 12 that should be possible. But most likely you have good reasons not to do that. 😉
